source/common/stats/symbol_table_impl.h

```diff
@@ -130,22 +70,94 @@ class SymbolEncoding {
+  /**
+   * Intermediate representation for a stat-name. This helps store multiple names
+   * in a single packed allocation. First we encode each desired name, then sum
+   * their sizes for the single packed allocation. This is used to store
+   * MetricImpl's tags and tagExtractedName. Like StatName, we don't want to pay
+   * a vptr overhead per object, and the representation is shared between the
+   * SymbolTable implementations, so this is just a pre-declare.
+   */
+  class Encoding {
+  public:
+    /**
+     * Before destructing SymbolEncoding, you must call moveToStorage. This
+     * transfers ownership, and in particular, the responsibility to call
+     * SymbolTable::clear() on all referenced symbols. If we ever wanted
+     * to be able to destruct a SymbolEncoding without transferring it
+     * we could add a clear(SymbolTable&) method.
+     */
+    ~Encoding();
+
+    /**
+     * Encodes a token into the vec.
+     *
+     * @param symbol the symbol to encode.
+     */
+    void addSymbol(Symbol symbol);
+
+    /**
+     * Encodes an entire string into the vec, on behalf of FakeSymbolTableImpl.
+     * TODO(jmarantz): delete this method when FakeSymbolTableImpl is deleted.
+     *
+     * @param str The string to encode.
+     */
+    void addStringForFakeSymbolTable(absl::string_view str);
+
+    /**
+     * Decodes a uint8_t array into a SymbolVec.
+     */
+    static SymbolVec decodeSymbols(const SymbolTable::Storage array, uint64_t size);
+
+    /**
+     * Returns the number of bytes required to represent StatName as a uint8_t
+     * array, including the encoded size.
+     */
+    uint64_t bytesRequired() const { return size() + StatNameSizeEncodingBytes; }
+
+    /**
+     * Returns the number of uint8_t entries we collected while adding symbols.
+     */
+    uint64_t size() const { return vec_.size(); }
+
+    /**
+     * Moves the contents of the vector into an allocated array. The array
+     * must have been allocated with bytesRequired() bytes.
+     *
+     * @param array destination memory to receive the encoded bytes.
+     * @return uint64_t the number of bytes transferred.
+     */
+    uint64_t moveToStorage(SymbolTable::Storage array);
+
+    void swap(Encoding& src) { vec_.swap(src.vec_); }
+
+  private:
+    std::vector<uint8_t> vec_;
+  };
+
   SymbolTableImpl();
   ~SymbolTableImpl() override;
 
   // SymbolTable
   std::string toString(const StatName& stat_name) const override;
-  SymbolEncoding encode(absl::string_view name) override;
   uint64_t numSymbols() const override;
   bool lessThan(const StatName& a, const StatName& b) const override;
   void free(const StatName& stat_name) override;
   void incRefCount(const StatName& stat_name) override;
   SymbolTable::StoragePtr join(const std::vector<StatName>& stat_names) const override;
 
+  void encode(absl::string_view name, Encoding& encoding);
+  void populateList(absl::string_view* names, int32_t num_names, StatNameList& list) override;
```

Comment:
Nit: group this, copyToBytes, callWithStringView under the rest of the private methods which override `SymbolTable`.